src/rta/fullsupernodelist.cpp

```diff
@@ -226,84 +229,64 @@ SupernodePtr FullSupernodeList::get(const string &address) const
     return SupernodePtr(nullptr);
 }
 
-bool FullSupernodeList::buildAuthSample(uint64_t height, vector<SupernodePtr> &out)
+bool FullSupernodeList::buildAuthSample(uint64_t height, const std::string& payment_id, vector<SupernodePtr> &out)
 {
-    crypto::hash block_hash;
-    string  block_hash_str;
-
-    MDEBUG("building auth sample for height: " << height);
+    boost::shared_lock<boost::shared_mutex> readerLock(m_access);
 
-    if (!getBlockHash(height - AUTH_SAMPLE_HASH_HEIGHT, block_hash_str)) {
-        LOG_ERROR("getBlockHash error");
+    if (height != m_blockchain_based_list_block_number)
+    {
+        LOG_ERROR("unable to build auth sample for block height " << height << " and PaymentID "
+          << payment_id << " because of mistmatch with blockchain based list block number " << m_blockchain_based_list_block_number);
         return false;
     }
 
-    epee::string_tools::hex_to_pod(block_hash_str, block_hash);
+    MDEBUG("building auth sample for height " << height << " and PaymentID " << payment_id);
 
-    std::array<std::vector<SupernodePtr>, TIERS> tier_supernodes;
+    out.clear();
+    out.reserve(AUTH_SAMPLE_SIZE);
+
+    const unsigned char* it               = reinterpret_cast<const unsigned char*>(payment_id.c_str());
```

Comment:
The payment_id is, I believe, a formatted UUID, which means the characters of this string are going to be a pretty bad source of entropy (each value of the first 7 is a hex digit; the 8th is an even hex digit; the 9th is always "-", the next 4 are very slow changing, etc.).

Reply:
Implementation has been updated according to UUID formatting rules.